File: src/server/orchestrator/runtimeStateWatcher.ts
import fs from 'node:fs';

export type DerivedRuntimeState = 'starting' | 'running' | 'waiting';

export interface RuntimeStateWatcherOptions {
  /** Path to the sentinel created by Claude Code's Stop hook. */
  waitingPath: string;
  /** Path to the sentinel created by Claude Code's UserPromptSubmit hook. */
  promptedPath: string;
  pollMs?: number;
  /** Fires on first poll AND on every change. */
  onChange: (state: DerivedRuntimeState) => void;
}

/**
 * Polls two sentinel files written by Claude Code hooks and derives a
 * runtime state. Fires `onChange` once on first poll (so reattach picks up
 * the current state) and again on every transition. Steady state is silent.
 *
 *   waiting present                  -> 'waiting'
 *   waiting absent, prompted present -> 'running'
 *   both absent                      -> 'starting'
 */
export class RuntimeStateWatcher {
  private timer: NodeJS.Timeout | null = null;
  private running = false;
  private last: DerivedRuntimeState | null = null;
  private readonly pollMs: number;

  constructor(private opts: RuntimeStateWatcherOptions) {
    this.pollMs = opts.pollMs ?? 500;
  }

  start(): void {
    if (this.running) return;
    this.running = true;
    // Fire the first poll synchronously so callers that immediately call
    // stop() still receive the initial state emission.
    try { this.readOnce(); } catch { /* best-effort */ }
    const tick = () => {
      if (!this.running) return;
      try { this.readOnce(); } catch { /* best-effort */ }
      this.timer = setTimeout(tick, this.pollMs);
    };
    this.timer = setTimeout(tick, this.pollMs);
  }

  stop(): void {
    this.running = false;
    if (this.timer) { clearTimeout(this.timer); this.timer = null; }
  }

  /** Synchronous poll — exposed for tests. */
  checkNow(): void { this.readOnce(); }

  private readOnce(): void {
    const waiting = fs.existsSync(this.opts.waitingPath);
    const prompted = fs.existsSync(this.opts.promptedPath);
    const derived: DerivedRuntimeState = waiting
      ? 'waiting'
      : prompted ? 'running' : 'starting';
    if (this.last === derived) return;
    this.last = derived;
    this.opts.onChange(derived);
  }
}

File: server-elixir/lib/fbi/orchestrator/runtime_state_watcher.ex
defmodule FBI.Orchestrator.RuntimeStateWatcher do
  @moduledoc "Port of src/server/orchestrator/runtimeStateWatcher.ts."
  use GenServer

  defstruct [:waiting_path, :prompted_path, :poll_ms, :on_change, :last, :timer]

  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts)
  end

  def stop(pid), do: GenServer.stop(pid)

  @impl true
  def init(opts) do
    state = %__MODULE__{
      waiting_path: Keyword.fetch!(opts, :waiting_path),
      prompted_path: Keyword.fetch!(opts, :prompted_path),
      poll_ms: Keyword.get(opts, :poll_ms, 500),
      on_change: Keyword.fetch!(opts, :on_change),
      last: nil,
      timer: nil
    }

    state = read_once(state)
    timer = Process.send_after(self(), :tick, state.poll_ms)
    {:ok, %{state | timer: timer}}
  end

  @impl true
  def handle_info(:tick, state) do
    state = read_once(state)
    timer = Process.send_after(self(), :tick, state.poll_ms)
    {:noreply, %{state | timer: timer}}
  end

  @impl true
  def terminate(_reason, state) do
    if state.timer, do: Process.cancel_timer(state.timer)
    :ok
  end

  defp read_once(state) do
    waiting = File.exists?(state.waiting_path)
    prompted = File.exists?(state.prompted_path)
    derived = if waiting, do: :waiting, else: if(prompted, do: :running, else: :starting)

    if derived != state.last do
      state.on_change.(derived)
      %{state | last: derived}
    else
      state
    end
  end
end
